11.0 mig stock picking package preparation

SimoRubi opened the pull request:
Standard migration

Commit:
Started stock_picking_pallet module

Allows to generate a pallet for a selection of pickings

Implements the generation of the package for the selected pickings

Add a sequence for the pallets

Add security for model stock.picking.pallet

Add the base files for the pallet report

Do no longer use a sequence for the name

Use the related pack name instead

Do not copy some fields on duplicate

Rename dest_pack_id to dest_package_id

Report: show infos and lines

Not really nicely formatted though. Some info are missing

Add weights and use a field for quants

Split in 2 steps: put in pack, set to done

The report can only be printed once it is done, because it needs the packs tree
to be created.

The life_date field depends on product_expiry

Improve layout

Add date_done when a pallet is done

Change conditions to hide fields

Rename dest_package_id to package_id

Rename the model the stock.picking.package.preparation

stock.picking.pallet was too specific, this module can be used
to generate any kind of package.

Rename files from pallet to package_preparation

Explicit the relations otherwise the name is too long

Remove the report: too specific

Add tests on preparation of the package

Add help messages along the way

Add a test on the weight

Improve the readme with the OCA's layout

Add Odoo Community Association (OCA) as author

Add translation pot

Add French translation

Allow the action_* methods to run on several records

Do not repeat the dict for the fields' states

yvaucher committed:
Add Bug tracker link

eLBati committed:
FIX when viewing weight fields when package does not exist

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

eLBati committed:
porting last changes from 8.0

Removing parts related to product.ul as this model does not exist anymore We can only compute net weight, because package weight does not exist

remove lot from operation tree, as it is a o2m now

using new method check_tracking

FIX tests

Adding my name as I ported modules to 9

udpate module version

stock_picking_package_preparation: IMP README and use short copyright header

Moving pack preparation under 'inventory control' as 'operations' is only available in debug mode

Porting translations from 8.0

stock_picking_package_preparation_line: porting translations form 8.0

IMP stock_picking_package_preparation: avoid @api.one

IMP stock_picking_package_preparation: use odoo noupdate

IMP stock_picking_package_preparation: use odoo noupdate
FIX PEP8

FIX stock_picking_package_preparation: compatibility with delivery and its ensure_one

copying stock_picking_package_preparation_line from 8.0

adapting to new values format from write method

FIX correctly handle write tuples

location_id and location_dest_id are mandatory in picking

using notebook as position based on string is not allowed

FIX tests

Adding my name as I ported modules to 9

udpate module version

stock_picking_package_preparation: IMP README and use short copyright header

pedrobaeza committed:
[MIG] Rename manifest files

archetipo committed:
[MIG] stock_picking_package_preparation: Migrated to 10.0

[MIG] stock_picking_package_preparation_line: Migrated to 10.0

[RM] view and model useless folder

tafaRU committed:
[stock_picking_package_preparation][FIX] 'select' is no longer supported; use 'index' instead

[stock_picking_package_preparation_lin][FIX] parameter 'digits_compute' is no longer supported; use 'digits' instead

[stock_picking_package_preparation] Move menu 'Package Preparations' under 'Operations'

[stock_picking_package_preparation] Update it.po

[FIX] stock_picking_package_preparation: remove comment

[IMP] stock_picking_package_preparation_line: improve error message for missing default location in picking type

OpenCode committed:
[IMP] Warning replace with UserError [FIX] Italian translation

oca-transbot committed:
OCA Transbot updated translations from Transifex

jado95 committed:
Date and Shipping Date set to 'copy=False'

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
remove obsolete .pot files [ci skip]

oca-travis committed:
[UPD] Update stock_picking_package_preparation.pot

SimoRubi committed:
[MIG] stock_picking_package_preparation: Migration to 11.0

manifest

Substitute stock.pack.operation with stock.move.line, after https://github.com/odoo/odoo/commit/ddf2b21bdf38492ce075d4ae96bb9323c6ef59dd

Removing children_ids after https://github.com/odoo/odoo/commit/809605bb8f16cf1398ff5070ec0d5bd568aef343

product packaging gets a variant instead of a template after https://github.com/odoo/odoo/commit/796900feb76f7ca5084d6b5f90cb24d5448cc326

get_content becomes get_contained_quants

test little refactoring

qty becomes quantity in stock.quant

[FIX] result_package_id has ondelete=restrict, since https://github.com/odoo/odoo/commit/809605bb8f16cf1398ff5070ec0d5bd568aef343

[FIX] update done qty only if it greater than 0

SimoRubi commented:
Tests still fail due to https://github.com/odoo/odoo/pull/26850

SimoRubi commented:
Now that https://github.com/odoo/odoo/pull/26850 is merged, @OCA/logistics-maintainers could you please review?
Thanks

nikul-serpentcs commented on stock_picking_package_preparation/__init__.py:
```diff
@@ -0,0 +1,5 @@
+# Author: Guewen Baconnier
```
@SimoRubi Remove Copyright from all  `__init__.py` file.

SimoRubi replied:
Thank for the review @nikul-serpentcs, done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/d922d5eb4af6d7d3d6cd338b8cc08ac180f3cb34

nikul-serpentcs commented on stock_picking_package_preparation/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+               '11.0/stock_picking_package_preparation',
+    'data': [
+        'views/stock_picking_package_preparation_view.xml',
+        'security/ir.model.access.csv',
```
@SimoRubi IMP `security`  file always first.

SimoRubi replied:
Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/d922d5eb4af6d7d3d6cd338b8cc08ac180f3cb34

nikul-serpentcs reviewed (commented):
Improve Code

SimoRubi commented:
@nikul-serpentcs is this ok for you now?

SimoRubi commented:
@nikul-serpentcs if this is ok for you, please do the review github-style :) thanks

nikul-serpentcs commented:
@SimoRubi Please attend `codecov` also, It's still False

tafaRU commented on stock_picking_package_preparation/README.rst:
```diff
@@ -0,0 +1,103 @@
+
+To use this module, you need to:
+
+ * Go to Warehouse > Package Preparation
```
s/Warehouse/Inventory

SimoRubi replied:
Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/3b10189d10ac3fa3e82aeeca990a61c6b566afbd

tafaRU reviewed (changes requested):
On runbot I created `PACK0000002`. Now if I click on the button *Cancel* I get the following error:

```
The operation cannot be completed, probably due to the following:
- deletion: you may be trying to delete a record while other records still reference it
- creation/update: a mandatory field is not correctly set

[object with reference: Packing Operation - stock.move.line]
```
Could you please check it?

tafaRU commented on stock_picking_package_preparation/readme/USAGE.rst:
```diff
@@ -0,0 +1,14 @@
+
+From there, you can create a new preparation.
+Choose a partner then the selection of pickings you want in the package.
+You can choose a logistic unit or a packaging, they will be set on the
```
How can I choose a logistic unit on the pack?

SimoRubi replied:
Logistic units existed in v8, as far as I know they were deleted in v10 already, anyway I've deleted that line in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/3b10189d10ac3fa3e82aeeca990a61c6b566afbd

eLBati reviewed (changes requested):
Tested steps:

 - Create a delivery order, add a new line and set quantity done = 1
 - Create a package preparation with the previously created picking
 - Click put in pack

System sets picking line quantity done to 0

rousseldenis commented on stock_picking_package_preparation/models/stock_picking_package_preparation.py:
```diff
@@ -0,0 +1,208 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api, _
+from odoo.exceptions import Warning as UserError
```
You can import UserError directly

SimoRubi replied:
Thanks for the review @rousseldenis, I missed that.
Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/169ab09cecbf3762a29dcb670bf29971fe120209

SimoRubi commented:
@tafaRU I fixed the error you described in https://github.com/OCA/stock-logistics-workflow/pull/448#pullrequestreview-159416061 with https://github.com/OCA/stock-logistics-workflow/pull/448/commits/3b10189d10ac3fa3e82aeeca990a61c6b566afbd

rousseldenis commented on stock_picking_package_preparation/models/stock_picking_package_preparation.py:
```diff
@@ -201,7 +201,9 @@ def _generate_pack(self):
-        for move_line in move_lines:
+        # If the stock.move.line has no processed quantity,
+        # set it equal to the initial demand
+        for move_line in move_lines.filtered(lambda ml: ml.qty_done == 0):
```
@SimoRubi Careful. On fields with precision, use float_compare()

SimoRubi replied:
Thanks, I didn't thought of that! Looking around a bit I found `float_is_zero` that is just what I need :) Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/169ab09cecbf3762a29dcb670bf29971fe120209

SimoRubi commented:
@eLBati I fixed your use-case in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/814764172396643c1004e1ac67a61fed5f03a04d and added the test `test_put_in_pack_keep_done_qty` to avoid the error to happen again.

The confusion is generated, in my opinion, from the different form presented when creating a picking in v10 vs v11:

- In v10 the only field that you were able to see was an anonymous *Quantity* (`product_uom_qty`)
- In v11 you can see both *Initial Demand* (`product_uom_qty`) and *Done quantity* (`quantity_done`)

Then setting only the *Done quantity* as you described in your use-case was not possible in v10 that's why that use-case was not covered **yet**

SimoRubi commented:
@nikul-serpentcs the coverage is already above 85% and I think I won't add more tests.
If you find use-cases that result in an error please tell me or add the tests you want creating a PR to my repository.

SimoRubi commented:
Thanks for your review @tafaRU, I have fixed [what you noticed](https://github.com/OCA/stock-logistics-workflow/pull/448#commitcomment-30685655) in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/18a90f1047417d727c615489403cf32a2a56afc1.
Is this ok for you now?

rousseldenis commented:
> @eLBati I fixed your use-case in [8147641](https://github.com/OCA/stock-logistics-workflow/commit/814764172396643c1004e1ac67a61fed5f03a04d) and added the test `test_put_in_pack_keep_done_qty` to avoid the error to happen again.
> 
> The confusion is generated, in my opinion, from the different form presented when creating a picking in v10 vs v11:
> 
> * In v10 the only field that you were able to see was an anonymous _Quantity_ (`product_uom_qty`)
> * In v11 you can see both _Initial Demand_ (`product_uom_qty`) and _Done quantity_ (`quantity_done`)
> 
> Then setting only the _Done quantity_ as you described in your use-case was not possible in v10 that's why that use-case was not covered **yet**

@SimoRubi In v10, the done quantity is present in pack operations.

rousseldenis commented on stock_picking_package_preparation/models/stock_picking_package_preparation.py:
```diff
@@ -0,0 +1,216 @@
+        for preparation in self:
+            package = preparation.package_id
+            if not package:
+                return
```
@SimoRubi I think this should be 'continue' instead as self can be a multi recordset

SimoRubi replied:
Thanks @rousseldenis, I missed that. Done in https://github.com/OCA/stock-logistics-workflow/pull/448/commits/20954520e6e2decb1728cada9ea7c7b14f583ccb

SimoRubi commented:
> > @eLBati I fixed your use-case in [8147641](https://github.com/OCA/stock-logistics-workflow/commit/814764172396643c1004e1ac67a61fed5f03a04d) and added the test `test_put_in_pack_keep_done_qty` to avoid the error to happen again.
> > The confusion is generated, in my opinion, from the different form presented when creating a picking in v10 vs v11:
> > 
> > * In v10 the only field that you were able to see was an anonymous _Quantity_ (`product_uom_qty`)
> > * In v11 you can see both _Initial Demand_ (`product_uom_qty`) and _Done quantity_ (`quantity_done`)
> > 
> > Then setting only the _Done quantity_ as you described in your use-case was not possible in v10 that's why that use-case was not covered **yet**
> 
> @SimoRubi In v10, the done quantity is present in pack operations.

@rousseldenis during the creation of a picking (it is still in draft state), in v10 only the tab **Initial Demand** is shown, and looks like there is no `quantity_done` (screenshot from runbot):
![image](https://user-images.githubusercontent.com/32064796/46313714-db68a700-c5c8-11e8-983b-de3e84889140.png)
Are you suggesting that its visibility is due to some configuration? Where can you see that field? Thanks

eLBati reviewed (approved):
Reviewed and tested

GSLabIt commented:
Any eta to be merged?

rousseldenis commented:
@tafaRU could you update your review?

rousseldenis reviewed (approved):
Code review

SimoRubi commented:
@joezsweet now that this is merged, I'll start working on https://github.com/OCA/stock-logistics-workflow/pull/447 but I can't give you an ETA as it depends on the OCA reviews.
